pkg/utils/networkinggadgets.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"os"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	containerutils "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils"
+)
+
+type NetworkingTracer[Event any] interface {
+	Attach(key string, pid uint32, eventCallback func(Event)) error
+	Detach(key string) error
+}
+
+func ConnectTracerToContainerCollection[Event any](
+	key string,
+	tracer NetworkingTracer[Event],
+	cc containercollection.ContainerResolver,
+	selector containercollection.ContainerSelector,
+	eventCallback func(Event),
+) error {
+	netnsHost, _ := containerutils.GetNetNs(os.Getpid())
+	genKey := func(container *containercollection.Container) string {
+		if container.Netns == netnsHost {
+			return "host"
+		}
+		return container.ID
+	}
+
+	attachContainerFunc := func(container *containercollection.Container) {
+		err := tracer.Attach(genKey(container), container.Pid, eventCallback)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: failed to trace container: %q: %s", container.Name, err)
```

Comment:
At the moment, unit tests expect the following messages:
```
t.publishMessage(trace, eventtypes.DEBUG, key, "tracer attached")
t.publishMessage(trace, eventtypes.ERR, key, fmt.Sprintf("failed to attach tracer: %s", err))
```
It seems too complex to pass additional callbacks for this.

What about passing a "publisher interface"  to ConnectNetworkingTracerToContainerCollection() that will implement `PublishEvent(tracerID string, line string) error`?

Reply:
> It seems too complex to pass additional callbacks for this.

Why do you think it's too complex? We can have an additional argument like `publishMessage func(eventtypes.EventType, *containercollection.Container, string),`

> At the moment, unit tests expect the following messages:

Do you think those messages are important? (I'm totally fine keeping them, I only want to understand their purpose).

Reply:
> > It seems too complex to pass additional callbacks for this.
> 
> Why do you think it's too complex? We can have an additional argument like `publishMessage func(eventtypes.EventType, *containercollection.Container, string),`

Ok :) I am fine with it.

I initially thought that would be too many arguments, so I tried to find another solution. But my solution is not necessarily smaller...

> > At the moment, unit tests expect the following messages:
> 
> Do you think those messages are important? (I'm totally fine keeping them, I only want to understand their purpose).

I think it's good to have them for users so they can find out if missing events are missing because they don't happen at all or if the container is not matching the container selector. But on the other hand we don't have that for non-network gadgets...